tests/clients/test_serveragent.py

```diff
@@ -0,0 +1,232 @@
+    "id": str(ENVIRONMENT_EXECUTION_STEP.id),
+    "command": ENVIRONMENT_EXECUTION_STEP.command,
+    "status": ENVIRONMENT_EXECUTION_STEP.status,
+    "startedAt": STARTED_AT_STRING,
+    "finishedAt": FINISHED_AT_STRING,
+}
+
+ENVIRONMENT_EXECUTION = EnvironmentExecution(
+    id=uuid.uuid4(), steps=[ENVIRONMENT_EXECUTION_STEP]
+)
+
+ENVIRONMENT_EXECUTION_BODY = {
+    "environmentId": str(ENVIRONMENT_EXECUTION.id),
+    "steps": [ENVIRONMENT_EXECUTION_STEP_BODY],
+}
+
+
+EXECUTION = Execution(
+    id=uuid.uuid4(),
+    status=ExecutionStatus.STARTED,
+    environments=[ENVIRONMENT_EXECUTION],
+    started_at=STARTED_AT,
+    finished_at=FINISHED_AT,
+)
+
+EXECUTION_BODY = {
+    "executionId": str(EXECUTION.id),
+    "status": EXECUTION.status,
+    "environments": [ENVIRONMENT_EXECUTION_BODY],
+    "startedAt": STARTED_AT_STRING,
+    "finishedAt": FINISHED_AT_STRING,
+}
+
+SSE_RESOURCE_EVENT = "@SSE/SERVER_RESOURCES_UPDATED"
+SSE_RESOURCE_MESSAGES = [
+    ServerSentEventMessage(
+        _id, SSE_RESOURCE_EVENT, json.dumps(SERVER_RESOURCES_BODY)
+    )
+    for _id in range(5)
+]
+
+SSE_LOG_EVENT = "log"
+SSE_LOG_MESSAGES = [
+    ServerSentEventMessage(
+        _id,
+        SSE_LOG_EVENT,
+        json.dumps([ENVIRONMENT_EXECUTION_STEP_LOG_LINE_BODY]),
+    )
+    for _id in range(5)
+]
+
+
+def test_cpu_usage_schema():
+    data = _CpuUsageSchema().load(CPU_USAGE_BODY)
+    assert data == CPU_USAGE
+
+
+def test_memory_usage_schema():
+    data = _MemoryUsageSchema().load(MEMORY_USAGE_BODY)
+    assert data == MEMORY_USAGE
+
+
+def test_server_resources_schema():
+    data = _ServerResourcesSchema().load(SERVER_RESOURCES_BODY)
+    assert data == SERVER_RESOURCES
+
+
+def test_environment_execution_step_log_schema():
+    data = _EnvironmentExecutionStepLogLineSchema().load(
+        ENVIRONMENT_EXECUTION_STEP_LOG_LINE_BODY
+    )
+    assert data == ENVIRONMENT_EXECUTION_STEP_LOG_LINE
+
+
+def test_environment_execution_step_schema():
+    data = _EnvironmentExecutionStepSchema().load(
+        ENVIRONMENT_EXECUTION_STEP_BODY
+    )
+    assert data == ENVIRONMENT_EXECUTION_STEP
+
+
+def test_environment_execution_schema():
+    data = _EnvironmentExecutionSchema().load(ENVIRONMENT_EXECUTION_BODY)
+    assert data == ENVIRONMENT_EXECUTION
+
+
+def test_execution_schema():
+    data = _ExecutionSchema().load(EXECUTION_BODY)
+    assert data == EXECUTION
+
+
+def test_client_latest_environment_execution(mocker):
+    mocker.patch.object(ServerAgentClient, "_get", return_value=EXECUTION)
+
+    client = ServerAgentClient(mocker.Mock(), mocker.Mock())
+    assert client.latest_environment_execution() == EXECUTION
+
+
+def test_client_stream_server_resources(mocker):
+    mocker.patch.object(
+        ServerAgentClient,
+        "_stream_server_sent_events",
+        return_value=SSE_RESOURCE_MESSAGES,
+    )
+
+    client = ServerAgentClient(mocker.Mock(), mocker.Mock())
+    for client_sse, mock_sse in zip(
```

Comment:
We discussed previously not using `zip` in case the two sequences are not the same length - do you think we can avoid this?